File: scripts/inspect-template.ts
/**
 * Script para inspeccionar la plantilla Excel FGL 059
 * Uso: npx tsx scripts/inspect-template.ts
 */
import ExcelJS from "exceljs";
import path from "path";

const TEMPLATE = path.resolve(
  "C:/Users/Usuario/Documents/ProyectoGrados/FGL 059 Monitoreo de Condiciones Ambientales del Laboratorio_V10 (7).xlsm"
);

async function main() {
  const wb = new ExcelJS.Workbook();
  try {
    await wb.xlsx.readFile(TEMPLATE);
  } catch (e) {
    console.error("Error leyendo archivo:", e);
    return;
  }

  console.log(`\nHojas en la plantilla (${wb.worksheets.length}):`);
  wb.worksheets.forEach((ws, i) => {
    console.log(`\n[${i + 1}] "${ws.name}" — ${ws.rowCount} filas, ${ws.columnCount} columnas`);

    // Primeras 40 filas de cada hoja
    let rowsPrinted = 0;
    ws.eachRow({ includeEmpty: false }, (row, rowNum) => {
      if (rowsPrinted >= 40) return;
      const vals: string[] = [];
      row.eachCell({ includeEmpty: true }, (cell, colNum) => {
        const v = cell.formula
          ? `[F:${cell.formula}]`
          : cell.value !== null && cell.value !== undefined
          ? String(cell.value)
          : "";
        if (v) vals.push(`C${colNum}="${v}"`);
      });
      if (vals.length) {
        console.log(`  R${rowNum}: ${vals.join(" | ")}`);
        rowsPrinted++;
      }
    });

    if (ws.rowCount > 40) {
      console.log(`  ... (${ws.rowCount - 40} filas más)`);
    }
  });
}

main();

File: scripts/inspect_template.py
import io
import msoffcrypto
import openpyxl

TEMPLATE = r"C:\Users\Usuario\Documents\ProyectoGrados\FGL 059 Monitoreo de Condiciones Ambientales del Laboratorio_V10 (7).xlsm"
PASSWORD = "Q1"

# Descifrar el archivo en memoria
with open(TEMPLATE, "rb") as f:
    office_file = msoffcrypto.OfficeFile(f)
    office_file.load_key(password=PASSWORD)
    decrypted = io.BytesIO()
    office_file.decrypt(decrypted)

decrypted.seek(0)
wb = openpyxl.load_workbook(decrypted, keep_vba=True, data_only=False)

print(f"\nHojas ({len(wb.sheetnames)}): {wb.sheetnames}\n")

for sheet_name in wb.sheetnames:
    ws = wb[sheet_name]
    print(f"\n{'='*60}")
    print(f"HOJA: \"{sheet_name}\" — {ws.max_row} filas x {ws.max_column} columnas")
    print(f"{'='*60}")

    printed = 0
    for row in ws.iter_rows():
        row_data = []
        for cell in row:
            val = None
            if cell.value is not None:
                val = str(cell.value)[:80]
            if val:
                row_data.append(f"  {cell.coordinate}: {val}")
        if row_data:
            print(f"Fila {row[0].row}:")
            for item in row_data:
                print(item)
            printed += 1
        if printed >= 60:
            print(f"  ... ({ws.max_row - row[0].row} filas más)")
            break

    # Mostrar celdas combinadas
    if ws.merged_cells:
        merges = list(str(m) for m in ws.merged_cells.ranges)[:20]
        print(f"\nCeldas combinadas: {merges}")
